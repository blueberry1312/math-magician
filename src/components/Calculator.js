import React, { useState } from 'react';
import './Calculator.css';
import calculate from '../logic/calculate';

export default function Calculator() {
  const [state, setState] = useState({
    total: null,
    next: null,
    operation: null,
  });

  const handleClick = (buttonName) => {
    const newState = calculate(state, buttonName);
    setState(newState);
  };

  return (
    <div className="container">
      <div className="result">{state.next || state.total || '0'}</div>
      <button type="button" className="btn-light btn-zero" onClick={() => handleClick('0')}>0</button>
      <button type="button" className="btn-light btn-dot" onClick={() => handleClick('.')}>.</button>
      <button type="button" className="btn-coloured btn-equal" onClick={() => handleClick('=')}>=</button>
      <button type="button" className="btn-light btn-clear" onClick={() => handleClick('AC')}>AC</button>
      <button type="button" className="btn-light btn-sign" onClick={() => handleClick('+/-')}>+/-</button>
      <button type="button" className="btn-light btn-percentage" onClick={() => handleClick('%')}>%</button>
      <button type="button" className="btn-coloured btn-divider" onClick={() => handleClick('รท')}>/</button>
      <button type="button" className="btn-coloured btn-multiply" onClick={() => handleClick('x')}>*</button>
      <button type="button" className="btn-coloured btn-substract" onClick={() => handleClick('-')}>-</button>
      <button type="button" className="btn-coloured btn-plus" onClick={() => handleClick('+')}>+</button>
      <button type="button" className="btn-light btn-one" onClick={() => handleClick('1')}>1</button>
      <button type="button" className="btn-light btn-two" onClick={() => handleClick('2')}>2</button>
      <button type="button" className="btn-light btn-three" onClick={() => handleClick('3')}>3</button>
      <button type="button" className="btn-light btn-four" onClick={() => handleClick('4')}>4</button>
      <button type="button" className="btn-light btn-five" onClick={() => handleClick('5')}>5</button>
      <button type="button" className="btn-light btn-six" onClick={() => handleClick('6')}>6</button>
      <button type="button" className="btn-light btn-seven" onClick={() => handleClick('7')}>7</button>
      <button type="button" className="btn-light btn-eight" onClick={() => handleClick('8')}>8</button>
      <button type="button" className="btn-light btn-nine" onClick={() => handleClick('9')}>9</button>
    </div>
  );
}
