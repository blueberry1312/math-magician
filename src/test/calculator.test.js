import { fireEvent, render, screen } from '@testing-library/react';
import Calculator from '../components/Calculator';

describe('Calculator test', () => {
  test('Calculator component is displayed on the HTML', () => {
    render(<Calculator />);
    const calculator = screen.getByTitle('calculator-container');
    expect(calculator).toMatchSnapshot();
  });

  test('Calculator buttons works and change the displayed result', () => {
    render(<Calculator />);
    const buttons = screen.getAllByRole('button');
    const result = screen.getByTitle('result');

    let cont = 1;
    buttons.forEach((button) => {
      fireEvent.click(button);
      switch (button.textContent) {
        case '.':
          expect(result.textContent).toBe(0 + button.textContent);
          break;
        case '=':
        case 'AC':
        case '+/-':
        case '%':
        case '+':
        case '/':
        case '-':
        case '*':
        case '0':
          expect(result.textContent).toBe('0');
          break;
        case '1':
          expect(result.textContent).toBe('1');
          break;
        default:
          expect(result.textContent.length).toBe(cont + 1);
          cont += 1;
          break;
      }
    });
  });
});
