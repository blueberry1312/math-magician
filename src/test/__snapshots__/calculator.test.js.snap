// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator test Calculator component is displayed on the HTML 1`] = `
<div
  class="container"
  title="calculator-container"
>
  <div
    class="result"
    title="result"
  >
    0
  </div>
  <button
    class="btn-light btn-zero"
    type="button"
  >
    0
  </button>
  <button
    class="btn-light btn-dot"
    type="button"
  >
    .
  </button>
  <button
    class="btn-coloured btn-equal"
    type="button"
  >
    =
  </button>
  <button
    class="btn-light btn-clear"
    type="button"
  >
    AC
  </button>
  <button
    class="btn-light btn-sign"
    type="button"
  >
    +/-
  </button>
  <button
    class="btn-light btn-percentage"
    type="button"
  >
    %
  </button>
  <button
    class="btn-coloured btn-divider"
    type="button"
  >
    /
  </button>
  <button
    class="btn-coloured btn-multiply"
    type="button"
  >
    *
  </button>
  <button
    class="btn-coloured btn-substract"
    type="button"
  >
    -
  </button>
  <button
    class="btn-coloured btn-plus"
    type="button"
  >
    +
  </button>
  <button
    class="btn-light btn-one"
    type="button"
  >
    1
  </button>
  <button
    class="btn-light btn-two"
    type="button"
  >
    2
  </button>
  <button
    class="btn-light btn-three"
    type="button"
  >
    3
  </button>
  <button
    class="btn-light btn-four"
    type="button"
  >
    4
  </button>
  <button
    class="btn-light btn-five"
    type="button"
  >
    5
  </button>
  <button
    class="btn-light btn-six"
    type="button"
  >
    6
  </button>
  <button
    class="btn-light btn-seven"
    type="button"
  >
    7
  </button>
  <button
    class="btn-light btn-eight"
    type="button"
  >
    8
  </button>
  <button
    class="btn-light btn-nine"
    type="button"
  >
    9
  </button>
</div>
`;
